### ISSUE TITLE
Detect "Functions are not valid as a React child." errors

### ISSUE THREAD

**Issue by donaldpipowitch** [Microsoft/TypeScript#26182]
## Search Terms

- "Functions are not valid as a React child."
- "Functions are not valid as a JSX child."

## Suggestion

In React it is not possible to return a function as a React child. It would be nice if TypeScript could catch this type error.

## Use Cases

Sometimes you want to use a function inside an JSX expression, but if you forget to call it, it will throw on runtime.

## Examples

See https://codesandbox.io/s/n3k4ooq3jj:

```ts
import * as React from "react";
import { render } from "react-dom";

const helloWorld = () => "Hello World!";

const App = () => <p>{helloWorld}</p>;
                            // ^---- should catch "Functions are not valid as a React child."
                            // ^---- {helloWorld()} would be correct

render(<App />, document.getElementById("root"));
```

## Checklist

My suggestion meets these guidelines:
* [ ] This wouldn't be a breaking change in existing TypeScript / JavaScript code ❓ It could break builds for invalid code, so... 🤷‍♀️ 
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. new expression-level syntax)

**Comment by RyanCavanaugh** [comment]
> In React it is not possible to return a function as a React child. 

We don't know what any given component does with its childen; this is only true for the built-in tags.

Presumably all we have to do is add a line to the react .d.ts that defines `children` of the intrinsic elements as
```ts
{
   (...args: any[])?: "Functions cannot be rendered";
}
```
and this should just work

**Comment by weswigham** [comment]
This is only allowed because `ReactNode` is poorly defined:
```ts
type ReactFragment = {} | ReactNodeArray;
type ReactNode = ReactChild | ReactFragment | ReactPortal | string | number | boolean | null | undefined;
```
Since `ReactFragment` contains `{}`, `ReactNode` is effectively `{} | null | undefined` which is effectively `unknown` or the top type, which can have anything assigned to it.

That `{}` should _probably_ be `object`.

**Comment by weswigham** [comment]
Yeeeeeah, I'm thinking that `ReactFragment` shouldn't contain `{}` at all - nowhere should that be required and it's a blanket licence to allow any type as a react child. (Plus, we typecheck fragments as type `JSX.Element` right now _anyway_)

**Comment by weswigham** [comment]
Looks like the React `0.13` .d.ts introduced this defect (https://github.com/DefinitelyTyped/DefinitelyTyped/pull/3813) in the types. 😦  There's no comment on why, so I imagine it was a workaround for something unexplained? It's probably safe to remove?

**Comment by domdomegg** [comment]
Linking some more relevant issues together:

https://github.com/DefinitelyTyped/DefinitelyTyped/issues/37596
https://github.com/DefinitelyTyped/DefinitelyTyped/issues/35622
https://github.com/DefinitelyTyped/DefinitelyTyped/issues/49705
https://github.com/DefinitelyTyped/DefinitelyTyped/issues/29307
